### ISSUE TITLE
Documentation for User{N} highlight groups is seriously lacking

### ISSUE THREAD

**Issue by lifecrisis** [vim/vim#1697]
A simple `:helpgrep User1` will return just five entries related to the special `User{N}` highlight groups.  The most extensive explanation of the use of these groups is found under `:h 'statusline'` an is quoted below:

```
	* -   Set highlight group to User{N}, where {N} is taken from the
	      minwid field, e.g. %1*.  Restore normal highlight with %* or %0*.
	      The difference between User{N} and StatusLine  will be applied
	      to StatusLineNC for the statusline of non-current windows.
	      The number N must be between 1 and 9.  See |hl-User1..9|
```

I stumbled upon this deficiency while attempting to use the special highlight groups in my own custom status line.

The documentation seems to indicate that the desired "darkening" effect for status lines in non-current windows will be achieved automatically when using a `User{N}` group in the status line.  However, this is not the case.  To have the status line refresh it's highlighting upon becoming inactive, you must use an autocommand.  This behavior seems more like a hack than a feature, and it doesn't seem to gel with the docs presented above.  It is also brittle, and is bound to fail when some other plugin uses the `'eventignore'` setting too liberally.

In summary, the documentation does not really explain the benefit of `User{N}` highlight groups. Why even have them if they don't provide a benefit over normal highlight groups? Are they intended to automatically be adjusted (as suggested in the quote above) to indicate buffer activity? 

I find the docs to be tremendously unclear on this subject and would be happy to help edit a revision if someone could enlighten me.

**Comment by lifecrisis** [comment]
As an additional reference, I would like to link to this [conversation on the vim mailing list](https://groups.google.com/forum/#!topic/vim_use/1IwOinGRLxA) that prompted me to submit this issue.

**Comment by brammool** [comment]
Jason Franklin wrote:

> A simple `:helpgrep User1` will return just five entries related to
> the special `User{N}` highlight groups.  The most extensive
> explanation of the use of these groups is found under `:h
> 'statusline'` an is quoted below:

Actually, the main part is ":help hl-User1".  Not sure what you are
expecting, but it's just the name of a highlight group, thus you can
assign colors.  There isn't really anything else to say.

> 
> ```
> 	* -   Set highlight group to User{N}, where {N} is taken from the
> 	      minwid field, e.g. %1*.  Restore normal highlight with %* or %0*.
> 	      The difference between User{N} and StatusLine  will be applied
> 	      to StatusLineNC for the statusline of non-current windows.
> 	      The number N must be between 1 and 9.  See |hl-User1..9|
> ```
> 
> I stumbled upon this deficiency while attempting to use the special
> highlight groups in my own custom status line.

The User1 - User9 groups are old, they were introduced as a shorthand.
Later the full names were added.

> The documentation seems to indicate that the desired "darkening"
> effect for status lines in non-current windows will be achieved
> automatically when using a `User{N}` group in the status line.
> However, this is not the case.  To have the status line refresh it's
> highlighting upon becoming inactive, you must use an autocommand.
> This behavior seems more like a hack than a feature, and it doesn't
> seem to gel with the docs presented above.

Take the docs literally: the difference is applied.

> In summary, the documentation does not really explain the benefit of
> 'User{N}' highlight groups. Why even have them if they don't provide a
> benefit over normal highlight groups? Are they intended to
> automatically be adjusted (as suggested in the quote above) to
> indicate buffer activity? 

It's just a name.

> I find the docs to be tremendously unclear on this subject and would
> be happy to help edit a revision if someone could enlighten me.

Perhaps a few examples would help.  But I think you are expecting too
much.

-- 
   [The rest of the ARMY stand around looking at a loss.]
INSPECTOR END OF FILM: (picks up megaphone) All right!  Clear off!  Go on!
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lifecrisis** [comment]
Hi, Bram!

I suppose I am over-playing the issue a bit, but I would like to try one more time to clarify my problem:

I am trying to design a `'statusline'` [prototype](https://en.wikipedia.org/wiki/Prototype-based_programming) that is extensible, functional, simple, and aesthetic.  Using this method, a master prototype can be cloned to create new status lines for special windows.  I've found that this cleans up the interface rapidly, and it would be a boon to Vim users if they could customize their `'statusline'` in an easier way.

My current design places all of the rendering logic within the master prototype, so that children of that prototype can be specified without defining an (often complicated) status line builder function. What I mean is that the following will create a new statusline:

```viml
let g:help_window_sl = g:StatusLine.New()
" ... specify attributes for the new help window status line here...
autocmd FileType help setlocal statusline=%!g:help_window_sl.render() 
```

If you're curious, my initial try at the idea is sitting in my [vimrc](https://github.com/lifecrisis/vim-vimrc/blob/master/vimrc#L434). Special statuslines have been defined for the following:

* normal text
* help buffers
* command window
* preview window
* quickfix window

The only problem I've had has been colors.  Status lines should "darken" in inactive windows and "lighten up" in active windows. From what I can tell, using either of the above methods of specifying a color in the status line leads to a *constant* color for that region.  The only way to modify this color in the expected way is to use an `autocommand`, which is bound to break eventually.

This leads to my question on the phrasing **"... the difference is applied... "**.  Does this mean that the expected "darkening" effect is made automatic? Or is this something I still have to do myself. I can't make heads or tails of this...

In another post, I mentioned the idea of a new status line item construct along the lines of `%#{ ... }#` where the ` ... ` is an expression returning a string that names a highlight group. This would solve the brittle highlighting problem once and for all. Of course, this is pointless if Vim already "darkens" windows for you when you use the proper construct (i.e. `User{N}` ...).  I hope this expresses the problem I'm having.  I want the "darkening" effect to be automatic and to NOT depend on autocommands. How can I do this?  Using the default groups (`StatusLine` and `StatusLineNC`) works as expected, but is not extensible to using more than two colors!

I don't want to seem lazy or demanding.  I love Vim and I actually want to contribute to Vim with plugins and with C code!  I'd be happy to modify the docs with some additional explanation. However, I'm still a young developer (only ~2 years in, 1yr with Vim) and my rate of code production is not as fast or as proficient as what you might see with a veteran.  One day though!

**Thank you** for your reply, Bram!!!

**Comment by rukbotto** [comment]
Regarding this issue I've found an interesting fact. I will further explain with an example.

Let's say you define a status line like this:

```vim
set statusline=
set statusline+=%1*
set statusline+=\ %f
```

Then you define a `StatusLine` and `StatusLineNC` highlight groups like this:

```vim
highlight! StatusLine ctermbg=114 ctermfg=237
highlight! StatusLineNC ctermbg=235 ctermfg=240
```

Finally you define the `User1` highlight group like this:

```vim
highlight! User1 ctermbg=114 ctermfg=237
```

Note that `User1` group has the same colors as `StatusLine` group.

This is the result:

![screen shot 2017-07-23 at 9 56 47 pm](https://user-images.githubusercontent.com/1110747/28506573-ef271fc0-6ff1-11e7-8296-61541d099e5d.png)

So using this config you could achieve the desired effect: Active windows will have the status line highlighted and inactive windows will have it dimmed automatically.

However, when `User1` is using colors different than those defined for `StatusLine` group, this effect is lost.

Hope this would clarify a bit the confusion around this topic.

**Comment by jsit** [comment]
@lifecrisis:

The way this was just explained to me helped me a lot:

> The difference between User{N} and StatusLine will be applied to StatusLineNC for the statusline of non-current windows.

Given:
```vim
hi StatusLine   cterm=reverse,bold ctermbg=black
hi User1        cterm=reverse,bold ctermbg=red
hi StatusLineNC cterm=reverse      ctermbg=black
```

The difference between `User1` and `StatusLine` is `ctermbg=red`.

Applying that difference to `StatusLineNC` would yield `cterm=reverse ctermbg=red`.

So an inactive statusline would display regions marked `%1*` as `cterm=reverse ctermbg=red` instead of `cterm=reverse,bold ctermbg=red`, which is what `User1` would normally display as.

Another way of saying it: `User1` will affect `StatusLineNC` only in the ways in which it also affects `StatusLine`.

**Comment by lifepillar** [comment]
>The only way to modify this color in the expected way is to use an `autocommand`

I also used to believe that, but other users in the past have pointed out to me that you never need to use autocommands: the status line knows when to update itself. To change colors, you may either dynamically change the definition of a highlight group according to some condition (by calling a function in `%{}` context), or use an idiom like `%#HiTrue#%{some_cond ? "text" : ""}%#HiFalse#%{some_cond ? "" : "alt_text"}%*`.

**Comment by lifecrisis** [comment]
@lifepillar, now that's useful.  I had never seen that idiom you mentioned before.  It looks like it would work very well.

**Comment by lifepillar** [comment]
@lifecrisis For an application of the other idea (dynamically update the highlight groups), you may take a look at the Status Line section in [my vimrc](https://github.com/lifepillar/vimrc/blob/master/vimrc). It's… dense, but self-contained.

(Now, @lifepillar supporting @lifecrisis: isn't that fun!)
